### PR TITLE
Add support for passing CA path (and file) when going through a proxy

### PR DESCRIPTION
**Pull request by CyberROFL**
The SDK allows to specify `caPath` (`CURLOPT_CAPATH`) and `caFile` (`CURLOPT_CAINFO`) options in `ClientConfiguration`, but these options are not enough when going through a proxy. There is a need to additionally specify `CURLOPT_PROXY_CAPATH` and `CURLOPT_PROXY_CAINFO`.  `proxyCaPath` and `proxyCaFile` are new options that allow to do this.

*Check all that applies:*
- [x] Did a review by yourself.
- [ ] Added proper tests to cover this PR. (If tests are not applicable, explain.)
- [ ] Checked if this PR is a breaking (APIs have been changed) change.
- [ ] Checked if this PR will _not_ introduce cross-platform inconsistent behavior.
- [ ] Checked if this PR would require a ReadMe/Wiki update.

Check which platforms you have built SDK on to verify the correctness of this PR.
- [x] Linux
- [x] Windows
- [ ] Android
- [ ] MacOS
- [ ] IOS
- [ ] Other Platforms

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by gab-david** [comment]
Hello,
We are also interested with this PR.
My team and I currently use a patch that modifies CurlHttpClient.cpp to reuse caPath and caFile.

```diff
diff --git a/aws-cpp-sdk-core/source/http/curl/CurlHttpClient.cpp b/aws-cpp-sdk-core/source/http/curl/CurlHttpClient.cpp
index 09b624c7dd..f8eca40fed 100644
--- a/aws-cpp-sdk-core/source/http/curl/CurlHttpClient.cpp
+++ b/aws-cpp-sdk-core/source/http/curl/CurlHttpClient.cpp
@@ -625,6 +625,15 @@ std::shared_ptr<HttpResponse> CurlHttpClient::MakeRequest(const std::shared_ptr<
                 curl_easy_setopt(connectionHandle, CURLOPT_PROXYUSERNAME, m_proxyUserName.c_str());
                 curl_easy_setopt(connectionHandle, CURLOPT_PROXYPASSWORD, m_proxyPassword.c_str());
             }
+            if(!m_caPath.empty())
+            {
+                curl_easy_setopt(connectionHandle, CURLOPT_PROXY_CAPATH, m_caPath.c_str());
+            }
+            if(!m_caFile.empty())
+            {
+                curl_easy_setopt(connectionHandle, CURLOPT_PROXY_CAINFO, m_caFile.c_str());
+            }
+
             curl_easy_setopt(connectionHandle, CURLOPT_NOPROXY, m_nonProxyHosts.c_str());
 #ifdef CURL_HAS_TLS_PROXY
             if (!m_proxySSLCertPath.empty())
```

This PR looks better than our patch.

Regards

**Comment by jmklix** [comment]
Could you rebase this change to resolve the merge conflicts?

**Comment by gab-david** [comment]
Rebase done in my fork https://github.com/aws/aws-sdk-cpp/pull/2318

**Review by DmitriyMusatkin — APPROVED** [review]

**Comment by CyberROFL** [comment]
@DmitriyMusatkin conflicts are resolved.

**Review by SergeyRyabinin — APPROVED** [review]

**Review by SergeyRyabinin — APPROVED** [review]

File: src/aws-cpp-sdk-core/source/http/curl/CurlHttpClient.cpp [hunk 1]
```diff
@@ -751,6 +751,14 @@ std::shared_ptr<HttpResponse> CurlHttpClient::MakeRequest(const std::shared_ptr<
             ss << m_proxyScheme << "://" << m_proxyHost;
             curl_easy_setopt(connectionHandle, CURLOPT_PROXY, ss.str().c_str());
             curl_easy_setopt(connectionHandle, CURLOPT_PROXYPORT, (long) m_proxyPort);
+            if(!m_proxyCaPath.empty())
+            {
+                curl_easy_setopt(connectionHandle, CURLOPT_PROXY_CAPATH, m_proxyCaPath.c_str());
```
Comment (sbiscigl) [inline src/aws-cpp-sdk-core/source/http/curl/CurlHttpClient.cpp:756]: looks like `CURLOPT_PROXY_CAPATH` and `CURLOPT_PROXY` were added in [7.52](https://curl.se/libcurl/c/CURLOPT_PROXY_CAPATH.html). we should likely ifdef guard this like we do the [ssl version](https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/source/http/curl/CurlHttpClient.cpp#L721). We have users that use versions [as low as 7.29](https://github.com/aws/aws-sdk-cpp/pull/2775)

**Review by sbiscigl — APPROVED** [review]

**Pull request merged**